### ISSUE TITLE
Regex handling

### ISSUE THREAD

**Issue by xogeny** [jsonata-js/jsonata#177]
It seems evaluating a "regex" node now returns a native function.  I'm not quite sure exactly what that is supposed to do.  As far as I can tell, the only place that a regular expression is discussed is as a potential argument to `$contains`, `$split`, `$match` and `$replace`.

So my question is mainly on the implementation side.  I'm not sure what it means to evaluate a regular expression by itself.  It seems like the only place if fits in is as an argument to some functions.  As such, it seems like it should actually just resolve to a literal that is a `Regexp` instance.  But in the current implementation, it resolves to a (native) function.

This would all be a matter of implementation details except that a) it is [not an error to evaluate a regexp by itself](http://try.jsonata.org/SJw8YUK9M) so we need to define what it should evaluate to and b) it is impacts what we pass into functions (besides the standard ones).

Am I missing some documentation?  It seems like the regular expression syntax isn't actually defined or mentioned anywhere in the docs (much less the semantics).  I would expect to something about the syntax perhaps find something under the documentation for [`/`](http://docs.jsonata.org/operators.html#-5).  But the only mention I found of regular expressions was for those standard functions and all the test cases involve invoking those functions.

Bottom line is that it seems like an implementation could just choose to treat a regular expression as a literal `Regexp` instance and it wouldn't really violate any of the semantics...as far as I can tell?

**Comment by deepilla** [comment]
A JSONata regex is a function that takes a string as an argument and returns an object that describes the leftmost match of the regex in the string, e.g.

```
/foo/ ("foobar")
```

returns

```
{
  "match": "foo",
  "start": 0,
  "end": 3,
  "groups": [],
  "next": "<native function>#0"
}
```

The item named "next" is a function that returns the next leftmost match. You can chain these calls together to progress through the matches.

```
/foo/ ("a string with lots of matches").next().next().next()
```

I don't think this behaviour is documented anywhere so I'm not certain of the idea behind it. But some of the JSONata tests use regexes in predicate expressions like this:

```
Account.Order.Product[`Product Name` ~> /hat/i].ProductID
```

I'm sure Andy can give you the background.

**Comment by andrew-coleman** [comment]
Exactly right.  To keep the JSONata type system simple, I have only added the 'function' type to the set of base JSON types (object, array, string, number, boolean, null).  So I modelled a regex as a function with a particular signature (param & return structure) rather than invent a new 'type'.  The regex syntax `/.../` is like a higher-order function that returns a 'matching function' as described by @deepilla above.  This matching function, when invoked on a string will return details of the match(es), or `undefined` for no-match.  This has the nice property that the expression in the predicate `` `Product Name` ~> /hat/i`` returns `true` if there is a match using the standard `~>` chaining operator.

The functions `$contains`, `$split`, `$match` and `$replace` have all been engineered to work with a function that returns an object of this shape regardless of whether that function was generated by the `/.../` regex syntax.  This has the benefit (although I never got round to documenting this - bad on me!) that users can create their own plug-in matching higher-order functions (e.g. `$glob`, `$soundsLike`, `$magicMatcher`), and provided those HOFs return functions that conform to the signature above, they can then be used as the match argument to the `$contains`, `$split`, `$match` and `$replace` functions.  In effect, these four functions are pluggable with different matching strategies, of which Regex is just one (built in) strategy.  I really must get round to documenting all this.

**Comment by xogeny** [comment]
OK, that all jives with what I'm seeing.  But there is something else going on as well...the signature for most of these functions (*e.g.,* `$replace`) show an optional final argument for `limit` that should be *a number*.  BUT, the test cases include *functions* for the limit.  It isn't clear to me how that is supposed to work (at least from looking at the code).

**Comment by xogeny** [comment]
Also, I'm going to add some test cases that test this specific functionality.

**Comment by andrew-coleman** [comment]
Which test case(s) are you referring to?  Certainly there are some test cases for `$replace` (e.g. [regex/case034](https://github.com/jsonata-js/jsonata/blob/master/test/test-suite/groups/regex/case034.json)) that have functions for the final argument, but this aligns with the `replacement` argument which can be either a [string or function](http://docs.jsonata.org/string-functions.html#replacestr-pattern-replacement--limit).  I would hope that if you supplied a function to the `limit` arg, you would get a signature validation error (T0410).
